### PR TITLE
Rebuild to enable `libcugraph`

### LINKED ISSUE #112: cuGraph dependency
There is some discussion going on about using RAPIDS cuGraph for some of the graph operations that will be add to CuPy (mirroring APIs of this in SciPy) ( https://github.com/cupy/cupy/issues/4219 ). One of the questions that comes out of this is how we handle the dependency for the `cupy` package. This depends a bit on how CuPy uses cuGraph (as such this is not totally flushed out). In any event am raising here for awareness/early discussion about how to handle this issue. Here are some thoughts on how this dependency gets managed:

1. Reuse Conda package for cuGraph from RAPIDS
2. Package cuGraph in conda-forge
3. Use a vendored cuGraph from CuPy (maybe CuPy uses a git submodule? maybe it gets cloned and built in the recipe?)
4. Make cuGraph an optional runtime dependency (using Python API, `dlopen`, or similar; this depends on how cuGraph is used by CuPy)
5. ?

These are very rough ideas and likely subject to change, but hopefully this starts the discussion. Should add the typical response to adding packages from other channels in conda-forge is no this is not an option. IOW 1 is not likely feasible, but have listed it for completeness anyways. 2 likely has challenges as well as there is an open question of who would maintain this and may lead to duplicated effort (RAPIDS will likely still build this in its own channel anyways). 3 would increase the complexity of the build here (and possibly for CuPy wheels depending on how it is done). 4 may increase the complexity of code added to CuPy. If others have different suggestions on how to handle these issue, that would be welcome

cc @rlratzel @pentchev @anaruse @kmaehashi @leofang

### PR DESCRIPTION
**Pull request by leofang**
Close #112.

TODO: do not use CF's `thrust` package, as it's untested. 

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by leofang** [comment]
cc: @jakirkham @kmaehashi @kkraus14 @rlratzel

**Comment by jakirkham** [comment]
Awesome Leo! 😄 Thank you for sticking with this 🙏 Looks like it is working thus far 🎉

Rick also did some more local testing using these changes yesterday, which worked well. We can also take these for another spin once we are ready to upload them 🙂

**Comment by leofang** [comment]
> Awesome Leo! 😄 Thank you for sticking with this 🙏 Looks like it is working thus far 🎉

np, though next time if I do this I hope I am getting paid somehow 😂

> Rick also did some more local testing using these changes yesterday, which worked well. We can also take these for another spin once we are ready to upload them 🙂

Would you like to keep the artifacts for local testing?

**Comment by leofang** [comment]
> TODO: do not use CF's `thrust` package, as it's untested.

Looks like we are safe --- the `thrust` package (v1.11.0) is not used at build time at all. I'd like to remove it from part of the build dependency, though. How should we approach this? Something like this?
```
ignore_run_exports:
  - thrust
```

**Comment by leofang** [comment]
Thanks, @rlratzel. Let's wait for the CI to finish. Note that because of https://github.com/conda-forge/conda-smithy/issues/1466, some CIs would fail due to filename collision of the artifacts, so be aware when you pick a particular build for testing.

**Comment by rlratzel** [comment]
@leofang - Good news, my local test of the artifact above passed. (thanks to @jakirkham for walking me through how to install the artifact from the CI report links)

I ran a [demo script](https://gist.github.com/rlratzel/161b3d5d5386311ee5f76bf780d38137) I wrote, which exercises the same example shown in the [SciPy docs for the same API](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csgraph.connected_components.html).

image

and it produced the expected output.

One observation though:
I noticed that `spdlog` and `thrust` were installed with `libcugraph`. I'm assuming this is because of the `run`/`run_exports` changes recently made to `librmm`, but I wonder if users will complain since `spdlog` isn't needed by `cupy`/`libcugraph` at runtime (and `thrust` isn't either?).

Also, my demo script includes a `print(graph)` statement just to match SciPy's example output. That statement causes `scipy` to be required, but when I comment the print out, `scipy` isn't needed.

**Comment by rlratzel** [comment]
> `spdlog` isn't needed by `cupy`/`libcugraph` at runtime (and `thrust` isn't either?).

well, for that matter, `librmm` isn't needed by `cupy/libcugraph` at runtime either, so I guess these are all problems resolved by fixing our leaked headers?

**Comment by jakirkham** [comment]
Yeah I think the dependency questions are something Leo is thinking about here ( https://github.com/conda-forge/cupy-feedstock/pull/126#issuecomment-828989608 )

**Comment by leofang** [comment]
Great, thanks for testing Rick!

>> I noticed that `spdlog` and `thrust` were installed with `libcugraph`. I'm assuming this is because of the `run`/`run_exports` changes recently made to `librmm`, but I wonder if users will complain since `spdlog` isn't needed by `cupy`/`libcugraph` at runtime (and `thrust` isn't either?).
>
> well, for that matter, `librmm` isn't needed by `cupy/libcugraph` at runtime either, so I guess these are all expected problems resolved by fixing our leaked headers?

For `librmm`, we need it at build time, but not runtime. For `thrust`, we don't even need it at build time (the docker image already has a copy, which CuPy uses). We should consider my earlier question https://github.com/conda-forge/cupy-feedstock/pull/126#issuecomment-828989608 (which is for @jakirkham 😅) to remove them.

>> Also, my demo script includes a `print(graph)` statement just to match SciPy's example output. That statement causes `scipy` to be required, but when I comment the print out, `scipy` isn't needed.

`scipy` is an optional dependency of `cupy`, so yes!

**Comment by leofang** [comment]
> Perhaps `ignore_run_exports_from` is more straightforward and fixes the `spdlog` point Rick mentions above. Also keeps `cupy` from learning to much about what's going on with `librmm`, which seems good from a simplification standpoint 🙂

Got it, that's a good point to fix two things at once @jakirkham! Will do shortly.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -66,9 +65,10 @@ build:
     {% if cuda_major == 11 %}
     - cusparselt  # [linux64]
     {% endif %}
-    #{% if cuda_major_minor >= (10, 1) %}
-    #- libcugraph  # [linux64]
-    #{% endif %}
+    {% if cuda_major_minor >= (10, 1) %}
+    - libcugraph  # [linux64]
```
Comment (jakirkham) [inline recipe/meta.yaml:69]: It appears `libcugraph` isn't included in the `run` requirements. Wonder if we might need to leave this out of this list to ensure it is included

Comment (leofang) [reply recipe/meta.yaml:69]: It is added in run_constrained as it's an optional dependency. If we decide to make cudnn optional to save disk usage, the same argument should be applied to libcugraph too as it's quite big.

**Review by leofang — COMMENTED** [review]
sod log and thrust are still installed during the build stage.  Looks like ignore_run_exports_from is not working?

**Comment by leofang** [comment]
> spdlog and thrust are still installed during the build stage. Looks like ignore_run_exports_from is not working?

Switched to `ignore_run_exports` in ada6532 and it doesn't work either...I suspect it's because `librmm` (where these two come from) is not a direct dependency.

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/meta.yaml:69]: Ah ok. Yeah that makes sense

**Comment by jakirkham** [comment]
Think it is ok that they are included in `host` as we need them to resolve the missing symbol issues that we had with `libcugraph` before.

The main thing with `ignore_run_exports`, etc. is to make sure that we are not getting those dependencies installed at runtime (IOW users are not needing to pick up those extra packages), which does not appear to be the case. So seems like this is working as expected (though feel free to correct me if I'm missing something else here 🙂)

**Comment by jakirkham** [comment]
All good. Seems we each got confused about different things 😅

Yep this is all good on our end. Thanks for working on this Leo! 🥳

**Comment by rlratzel** [comment]
Thanks @leofang and @jakirkham !

**Pull request merged**